src/v/cluster/scheduling/constraints.cc

```diff
@@ -190,4 +260,69 @@ soft_constraint_evaluator distinct_rack(
     return soft_constraint_evaluator(std::make_unique<impl>(replicas, state));
 }
 
+soft_constraint_evaluator least_disk_filled(
+  const cluster_health_report& health_report,
+  const float& max_disc_usage,
+  const std::unordered_map<model::node_id, uint64_t>&
+    assigned_reallocation_sizes) {
+    class impl : public soft_constraint_evaluator::impl {
+    public:
+        impl(
+          const cluster_health_report& health_report,
+          const float& max_disc_usage,
+          const std::unordered_map<model::node_id, uint64_t>&
+            assigned_reallocation_sizes)
+          : _health_report(health_report)
+          , _max_disc_usage(max_disc_usage)
+          , _assigned_reallocation_sizes(assigned_reallocation_sizes) {}
+        uint64_t score(const allocation_node& node) const final {
+            // we return 0 for fully filled node and 10'000'000 for nodes
+            // with disk capacity available
+            auto equal_node_id =
+              [&node](const node_health_report& health_report) {
+                  return health_report.id == node.id();
+              };
+            auto node_health = std::find_if(
+              _health_report.node_reports.begin(),
+              _health_report.node_reports.end(),
+              equal_node_id);
+            if (node_health == _health_report.node_reports.end()) {
+                return 0;
+            } else {
+                uint64_t assigned_reallocation_size = 0;
+                if (_assigned_reallocation_sizes.contains(node.id())) {
+                    assigned_reallocation_size
+                      = _assigned_reallocation_sizes.at(node.id());
+                }
+                for (const auto& disk : node_health->local_state.disks) {
+                    auto disk_free_space_percent
+                      = double(disk.free - assigned_reallocation_size)
+                        / double(disk.total);
+                    auto min_disk_free_space = double(1) - _max_disc_usage;
```

Comment:
nit: probably should just use doubles everywhere

Reply:
I think that float is enough for this.
Forgot chaning to float here